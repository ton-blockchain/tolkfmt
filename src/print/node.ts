import type {Node} from "web-tree-sitter"
import type {Ctx} from "./ctx"
import type {Doc} from "../doc"
import {concat, hardLine, text} from "../doc"
import * as stmts from "./stmts"
import * as decls from "./decls"
import * as expr from "./expr"
import * as types from "./types"
import type {CommentInfo} from "../comments"
import {takeTrailing, takeLeading} from "../comments"
import type {Range} from "../index"

export const printNode = (node: Node, ctx: Ctx): Doc | undefined => {
    if (ctx.range && !nodeIntersectsRange(node, ctx.range)) {
        if (node.type !== "source_file") {
            return printOriginalNodeText(node, ctx)
        }
    }

    if (node.type === "source_file") {
        return decls.printSourceFile(node, ctx)
    }

    if (node.type === "number_literal") {
        return expr.printNumberLiteral(node, ctx)
    }

    if (node.type === "string_literal") {
        return expr.printStringLiteral(node, ctx)
    }

    if (node.type === "boolean_literal") {
        return expr.printBooleanLiteral(node, ctx)
    }

    if (node.type === "null_literal") {
        return expr.printNullLiteral(node, ctx)
    }

    if (node.type === "underscore") {
        return expr.printUnderscore(node, ctx)
    }

    if (node.type === "identifier" || node.type === "type_identifier") {
        return expr.printIdentifier(node, ctx)
    }

    if (node.type === "type_alias_declaration") {
        return decls.printTypeAlias(node, ctx)
    }

    if (node.type === "constant_declaration") {
        return decls.printConstantDeclaration(node, ctx)
    }

    if (node.type === "function_declaration") {
        return decls.printFunction(node, ctx)
    }

    if (node.type === "union_type") {
        return types.printUnionType(node, ctx)
    }

    if (node.type === "nullable_type") {
        return types.printNullableType(node, ctx)
    }

    if (node.type === "parenthesized_type") {
        return types.printParenthesizedType(node, ctx)
    }

    if (node.type === "tensor_type") {
        return types.printTensorType(node, ctx)
    }

    if (node.type === "tuple_type") {
        return types.printTupleType(node, ctx)
    }

    if (node.type === "dot_access") {
        return expr.printDotAccess(node, ctx)
    }

    if (node.type === "function_call") {
        return expr.printFunctionCall(node, ctx)
    }

    if (node.type === "argument_list") {
        return expr.printArgumentList(node, ctx)
    }

    if (node.type === "call_argument") {
        return expr.printCallArgument(node, ctx)
    }

    if (node.type === "binary_operator") {
        return expr.printBinaryExpression(node, ctx)
    }

    if (node.type === "assignment") {
        return stmts.printAssignment(node, ctx)
    }

    if (node.type === "unary_operator") {
        return expr.printUnaryOperator(node, ctx)
    }

    if (node.type === "parenthesized_expression") {
        return expr.printParenthesizedExpression(node, ctx)
    }

    if (node.type === "tensor_expression") {
        return expr.printTensorExpression(node, ctx)
    }

    if (node.type === "typed_tuple") {
        return expr.printTypedTuple(node, ctx)
    }

    if (node.type === "cast_as_operator") {
        return expr.printCastAsOperator(node, ctx)
    }

    if (node.type === "is_type_operator") {
        return expr.printIsTypeOperator(node, ctx)
    }

    if (node.type === "not_null_operator") {
        return expr.printNotNullOperator(node, ctx)
    }

    if (node.type === "lazy_expression") {
        return expr.printLazyExpression(node, ctx)
    }

    if (node.type === "ternary_operator") {
        return expr.printTernaryOperator(node, ctx)
    }

    if (node.type === "object_literal") {
        return expr.printObjectLiteral(node, ctx)
    }

    if (node.type === "object_literal_body") {
        return expr.printObjectLiteralBody(node, ctx)
    }

    if (node.type === "instance_argument") {
        return expr.printInstanceArgument(node, ctx)
    }

    if (node.type === "if_statement") {
        return stmts.printIfStatement(node, ctx)
    }

    if (node.type === "block_statement") {
        return stmts.printBlockStatement(node, ctx)
    }

    if (node.type === "expression_statement") {
        return stmts.printExpressionStatement(node, ctx)
    }

    if (node.type === "return_statement") {
        return stmts.printReturnStatement(node, ctx)
    }

    if (node.type === "break_statement") {
        return stmts.printBreakStatement(node, ctx)
    }

    if (node.type === "continue_statement") {
        return stmts.printContinueStatement(node, ctx)
    }

    if (node.type === "throw_statement") {
        return stmts.printThrowStatement(node, ctx)
    }

    if (node.type === "while_statement") {
        return stmts.printWhileStatement(node, ctx)
    }

    if (node.type === "do_while_statement") {
        return stmts.printDoWhileStatement(node, ctx)
    }

    if (node.type === "repeat_statement") {
        return stmts.printRepeatStatement(node, ctx)
    }

    if (node.type === "local_vars_declaration") {
        return stmts.printLocalVarsDeclaration(node, ctx)
    }

    if (node.type === "var_declaration") {
        return stmts.printVarDeclaration(node, ctx)
    }

    if (node.type === "tuple_vars_declaration") {
        return stmts.printTupleVarsDeclaration(node, ctx)
    }

    if (node.type === "tensor_vars_declaration") {
        return stmts.printTensorVarsDeclaration(node, ctx)
    }

    if (node.type === "parameter_list") {
        return decls.printParameterList(node, ctx)
    }

    if (node.type === "parameter_declaration") {
        return decls.printParameterDeclaration(node, ctx)
    }

    if (node.type === "method_declaration") {
        return decls.printMethodDeclaration(node, ctx)
    }

    if (node.type === "get_method_declaration") {
        return decls.printGetMethodDeclaration(node, ctx)
    }

    // Top-level declarations
    if (node.type === "tolk_required_version") {
        return decls.printTolkRequiredVersion(node, ctx)
    }

    if (node.type === "version_value") {
        return decls.printVersionValue(node, ctx)
    }

    if (node.type === "import_directive") {
        return decls.printImportDirective(node, ctx)
    }

    if (node.type === "global_var_declaration") {
        return decls.printGlobalVarDeclaration(node, ctx)
    }

    if (node.type === "struct_declaration") {
        return decls.printStructDeclaration(node, ctx)
    }

    if (node.type === "struct_body") {
        return decls.printStructBody(node, ctx)
    }

    if (node.type === "struct_field_declaration") {
        return decls.printStructFieldDeclaration(node, ctx)
    }

    if (node.type === "empty_statement") {
        return stmts.printEmptyStatement(node, ctx)
    }

    if (node.type === "set_assignment") {
        return stmts.printSetAssignment(node, ctx)
    }

    if (node.type === "type_parameters") {
        return decls.printTypeParameters(node, ctx)
    }

    if (node.type === "type_parameter") {
        return decls.printTypeParameter(node, ctx)
    }

    if (node.type === "type_instantiatedTs") {
        return expr.printTypeInstantiatedTs(node, ctx)
    }

    if (node.type === "generic_instantiation") {
        return expr.printGenericInstantiation(node, ctx)
    }

    if (node.type === "instantiationT_list") {
        return expr.printInstantiationTList(node, ctx)
    }

    if (node.type === "asm_body") {
        return decls.printAsmBody(node, ctx)
    }

    if (node.type === "builtin_specifier") {
        return decls.printBuiltinSpecifier(node, ctx)
    }

    if (node.type === "method_receiver") {
        return decls.printMethodReceiver(node, ctx)
    }

    if (node.type === "annotation_list") {
        return decls.printAnnotationList(node, ctx)
    }

    if (node.type === "annotation") {
        return decls.printAnnotation(node, ctx)
    }

    if (node.type === "annotation_arguments") {
        return decls.printAnnotationArguments(node, ctx)
    }

    if (node.type === "fun_callable_type") {
        return types.printFunCallableType(node, ctx)
    }

    if (node.type === "assert_statement") {
        return stmts.printAssertStatement(node, ctx)
    }

    if (node.type === "try_catch_statement") {
        return stmts.printTryCatchStatement(node, ctx)
    }

    if (node.type === "catch_clause") {
        return stmts.printCatchClause(node, ctx)
    }

    if (node.type === "match_statement") {
        return stmts.printMatchStatement(node, ctx)
    }

    if (node.type === "match_expression") {
        return expr.printMatchExpression(node, ctx)
    }

    if (node.type === "match_body") {
        return expr.printMatchBody(node, ctx)
    }

    if (node.type === "match_arm") {
        return expr.printMatchArm(node, ctx)
    }

    if (node.type === "numeric_index") {
        return expr.printNumericIndex(node, ctx)
    }

    return undefined
}

export function formatLeading(leading: CommentInfo[]): Doc[] {
    if (leading.length === 1 && leading[0].text.startsWith("/*")) {
        return [text(leading[0].text)] // inline /* foo */
    }

    return leading.flatMap(c => [text(c.text), hardLine()])
}

export function formatDangling(dangling: CommentInfo[]): Doc[] {
    if (dangling.length === 0) {
        return []
    }

    if (dangling.length === 1) {
        return [text(dangling[0].text)]
    }

    const [first, ...rest] = dangling
    return [text(first.text), ...rest.flatMap(c => [hardLine(), text(c.text)])]
}

function nodeIntersectsRange(node: Node, range: Range): boolean {
    const nodeStart = node.startPosition
    const nodeEnd = node.endPosition

    if (
        nodeEnd.row < range.start.line ||
        (nodeEnd.row === range.start.line && nodeEnd.column < range.start.character)
    ) {
        return false
    }

    if (
        nodeStart.row > range.end.line ||
        (nodeStart.row === range.end.line && nodeStart.column > range.end.character)
    ) {
        return false
    }

    // intersects
    return true
}

export function getOriginalNodeText(node: Node): string {
    // semicolon is not a part of the following nodes, so we need to add it manually
    const needSemicolon =
        node.type === "local_vars_declaration" ||
        node.type === "return_statement" ||
        node.type === "do_while_statement" ||
        node.type === "break_statement" ||
        node.type === "continue_statement" ||
        node.type === "throw_statement" ||
        node.type === "assert_statement" ||
        node.type === "expression_statement"

    return node.text + (needSemicolon ? ";" : "")
}

export function printOriginalNodeText(node: Node, ctx: Ctx): Doc {
    const leading = takeLeading(node, ctx.comments)
    const trailing = takeTrailing(node, ctx.comments)
    const leadingDoc = leading.map(c => concat([text(c.text), hardLine()]))
    const trailingDoc = trailing.map(c => concat([text(" "), text(c.text)]))
    return concat([...leadingDoc, text(getOriginalNodeText(node)), ...trailingDoc])
}

export function isFmtIgnoreComment(comment: CommentInfo): boolean {
    return comment.text.trim() === "// fmt-ignore"
}

export function hasFmtIgnoreDirective(leadingComments: CommentInfo[]): boolean {
    return leadingComments.some(comment => isFmtIgnoreComment(comment))
}
