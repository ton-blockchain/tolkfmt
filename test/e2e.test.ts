import {join, normalize} from "node:path"
import type {CommandResult} from "./utils/test-util"
import {runCommand} from "./utils/test-util"
import {readFileSync, rmSync, writeFileSync} from "node:fs"
import {mkdir} from "node:fs/promises"

// disable tests on windows
const testExceptWindows =
    process.platform === "win32" && Boolean(process.env["CI"]) ? test.skip : test

const tolkfmt = async (...args: string[]): Promise<CommandResult> => {
    const tolkfmtPath = normalize(join(__dirname, "..", "bin", "tolkfmt"))
    const command = `node ${tolkfmtPath} ${args.join(" ")}`
    return runCommand(command)
}

const outputDir = join(__dirname, "output")

const goodContract = `
fun test(x: int, y: string): string {
    return x + y
}
`

const badContract = `
fun   test(  x: int,y:string   ): string{
return    x+y;
}
`

const contractWithSyntaxError = `
fun test(x: int, y: string): string {
    return foo("hello world";
}
`

const structCode = `
struct MyStruct {
    x: int = 42,
    y: string = "hello"
}
`

describe("tolkfmt foo.tolk", () => {
    testExceptWindows("Exits with correct code", async () => {
        await mkdir(outputDir, {recursive: true})
        const file = join(outputDir, "contract.tolk")
        writeFileSync(file, goodContract)
        const result = await tolkfmt(file)
        expect(result).toMatchObject({kind: "exited", code: 0})

        rmSync(file)
    })

    testExceptWindows("Default run", async () => {
        await mkdir(outputDir, {recursive: true})
        const file = join(outputDir, "contract.tolk")
        writeFileSync(file, goodContract)
        const result = await tolkfmt(file)
        expect(result).toMatchSnapshot()

        rmSync(file)
    })

    testExceptWindows("Default run with write to file", async () => {
        await mkdir(outputDir, {recursive: true})
        const file = join(outputDir, "contract.tolk")
        writeFileSync(file, badContract)
        await tolkfmt(file, "-w")

        const newContent = readFileSync(file, "utf8")
        expect(newContent).toMatchSnapshot()

        rmSync(file)
    })

    testExceptWindows("Run on directory", async () => {
        const dir = outputDir
        const innerDir = join(dir, "inner")
        const innerInnerDir = join(innerDir, "inner-2")

        await mkdir(dir, {recursive: true})
        await mkdir(innerDir, {recursive: true})
        await mkdir(innerInnerDir, {recursive: true})

        // inner
        //   file1.tolk
        //   inner-2
        //      file2.tolk
        //      file3.tolk
        const file1 = join(innerDir, "file1.tolk")
        const file2 = join(innerInnerDir, "file2.tolk")
        const file3 = join(innerInnerDir, "file3.tolk")

        writeFileSync(file1, "fun foo1() {   }")
        writeFileSync(file2, "fun foo2() {  }")
        writeFileSync(file3, "fun foo3() {     }")

        await tolkfmt(innerDir, "-w")

        expect(readFileSync(file1, "utf8")).toMatchSnapshot()
        expect(readFileSync(file2, "utf8")).toMatchSnapshot()
        expect(readFileSync(file3, "utf8")).toMatchSnapshot()

        rmSync(innerDir, {recursive: true})
    })

    testExceptWindows("Check on directory with formatted files", async () => {
        const dir = outputDir
        const innerDir = join(dir, "inner")
        const innerInnerDir = join(innerDir, "inner-2")

        await mkdir(dir, {recursive: true})
        await mkdir(innerDir, {recursive: true})
        await mkdir(innerInnerDir, {recursive: true})

        const file1 = join(innerDir, "file1.tolk")
        const file2 = join(innerInnerDir, "file2.tolk")
        const file3 = join(innerInnerDir, "file3.tolk")

        writeFileSync(file1, "fun foo1() {}\n")
        writeFileSync(file2, "fun foo2() {}\n")
        writeFileSync(file3, "fun foo3() {}\n")

        const result = await tolkfmt(innerDir, "--check")
        expect(result).toMatchSnapshot()

        rmSync(innerDir, {recursive: true})
    })

    testExceptWindows("Check on directory with not formatted files", async () => {
        const dir = outputDir
        const innerDir = join(dir, "inner")
        const innerInnerDir = join(innerDir, "inner-2")

        await mkdir(dir, {recursive: true})
        await mkdir(innerDir, {recursive: true})
        await mkdir(innerInnerDir, {recursive: true})

        const file1 = join(innerDir, "file1.tolk")
        const file2 = join(innerInnerDir, "file2.tolk")
        const file3 = join(innerInnerDir, "file3.tolk")

        writeFileSync(file1, "fun foo1() {  }\n")
        writeFileSync(file2, "fun foo2() {  }\n")
        writeFileSync(file3, "fun foo3() {  }\n")

        const result = await tolkfmt(innerDir, "--check")
        expect(result).toMatchSnapshot()

        rmSync(innerDir, {recursive: true})
    })

    testExceptWindows("Check on several formatted files", async () => {
        const dir = outputDir
        const innerDir = join(dir, "inner")
        const innerInnerDir = join(innerDir, "inner-2")

        await mkdir(dir, {recursive: true})
        await mkdir(innerDir, {recursive: true})
        await mkdir(innerInnerDir, {recursive: true})

        const file1 = join(innerDir, "file1.tolk")
        const file2 = join(innerInnerDir, "file2.tolk")
        const file3 = join(innerInnerDir, "file3.tolk")

        writeFileSync(file1, "fun foo1() {}\n")
        writeFileSync(file2, "fun foo2() {  }\n") // not checked
        writeFileSync(file3, "fun foo3() {}\n")

        const result = await tolkfmt("--check", file1, file3)
        expect(result).toMatchSnapshot()

        rmSync(innerDir, {recursive: true})
    })

    testExceptWindows("Check on several directories", async () => {
        const dir = outputDir
        const innerDir = join(dir, "inner")
        const innerInnerDir = join(innerDir, "inner-2")
        const innerInnerDir2 = join(innerDir, "inner-3")

        await mkdir(dir, {recursive: true})
        await mkdir(innerDir, {recursive: true})
        await mkdir(innerInnerDir, {recursive: true})
        await mkdir(innerInnerDir2, {recursive: true})

        const file1 = join(innerDir, "file1.tolk")
        const file2 = join(innerInnerDir, "file2.tolk")
        const file3 = join(innerInnerDir2, "file3.tolk")

        writeFileSync(file1, "fun foo1() { }\n") // not checked
        writeFileSync(file2, "fun foo2() {  }\n")
        writeFileSync(file3, "fun foo3() {}\n")

        const result = await tolkfmt("--check", innerInnerDir, innerInnerDir2)
        expect(result).toMatchSnapshot()

        rmSync(innerDir, {recursive: true})
    })

    testExceptWindows("With syntax error", async () => {
        await mkdir(outputDir, {recursive: true})
        const file = join(outputDir, "contract.tolk")
        writeFileSync(file, contractWithSyntaxError)
        const result = await tolkfmt(file, "-w")

        // Check status without timing details
        expect(result.kind).toBe("exited")
        if (result.kind === "exited") {
            expect(result.code).toBe(0)
            expect(result.stderr).toBe("")
            expect(result.stdout).toMatch(/contract\.tolk \d+ms \(reformatted\)/)
        }

        rmSync(file)
    })

    testExceptWindows("Check and write flags simultaneously", async () => {
        await mkdir(outputDir, {recursive: true})
        const file = join(outputDir, "contract.tolk")
        writeFileSync(file, badContract)
        const result = await tolkfmt(file, "-w", "--check")
        expect(result).toMatchSnapshot()

        rmSync(file)
    })

    testExceptWindows("Format struct", async () => {
        await mkdir(outputDir, {recursive: true})
        const file = join(outputDir, "struct.tolk")
        writeFileSync(file, structCode)
        const result = await tolkfmt(file)
        expect(result).toMatchSnapshot()

        rmSync(file)
    })

    testExceptWindows("Version flag", async () => {
        const result = await tolkfmt("--version")
        if (result.kind === "exited") {
            const {stdout} = result
            expect(stdout.split(" ")[0]).toMatchSnapshot()
        }
    })

    testExceptWindows("Help flag", async () => {
        const result = await tolkfmt("--help")
        expect(result).toMatchSnapshot()
    })

    testExceptWindows("No arguments", async () => {
        const result = await tolkfmt()
        expect(result).toMatchSnapshot()
    })

    testExceptWindows("Range formatting: basic", async () => {
        await mkdir(outputDir, {recursive: true})
        const file = join(outputDir, "range-test.tolk")
        const codeWithBadFormatting = `fun foo(){
    val   x   =   1;
    val   y   =   2;
    val   z   =   3;
    return x + y + z;
}`
        writeFileSync(file, codeWithBadFormatting)

        // Format only the second line (val y = 2)
        const result = await tolkfmt(file, "--range", "1:4-1:20")
        expect(result).toMatchSnapshot()

        rmSync(file)
    })

    testExceptWindows("Range formatting: with check", async () => {
        await mkdir(outputDir, {recursive: true})
        const file = join(outputDir, "range-check.tolk")
        const codeWithBadFormatting = `fun foo(){
    val   x   =   1;
    val   y   =   2;
    val   z   =   3;
    return x + y + z;
}`
        writeFileSync(file, codeWithBadFormatting)

        // Check only the second line (not formatted)
        const result = await tolkfmt(file, "--range", "1:4-1:20", "--check")
        expect(result).toMatchSnapshot()

        rmSync(file)
    })

    testExceptWindows("Range formatting: multiple lines", async () => {
        await mkdir(outputDir, {recursive: true})
        const file = join(outputDir, "range-multi.tolk")
        const codeWithBadFormatting = `fun foo(){
    val   x   =   1;
    val   y   =   2;
    val   z   =   3;
    return x + y + z;
}`
        writeFileSync(file, codeWithBadFormatting)

        // Format lines 1-2 (val x and val y)
        const result = await tolkfmt(file, "--range", "1:0-2:20")
        expect(result).toMatchSnapshot()

        rmSync(file)
    })

    testExceptWindows("Range formatting: struct", async () => {
        await mkdir(outputDir, {recursive: true})
        const file = join(outputDir, "range-struct.tolk")
        const codeWithBadFormatting = `type MyType = int;

struct MyStruct {
    field1:   int;
    field2:   string;
}

fun foo() {
    val   x   =   1;
}`
        writeFileSync(file, codeWithBadFormatting)

        // Format only the struct
        const result = await tolkfmt(file, "--range", "2:0-5:1")
        expect(result).toMatchSnapshot()

        rmSync(file)
    })

    testExceptWindows("Range formatting: invalid format", async () => {
        await mkdir(outputDir, {recursive: true})
        const file = join(outputDir, "range-invalid.tolk")
        writeFileSync(file, goodContract)

        // Invalid range format
        const result = await tolkfmt(file, "--range", "invalid-range")
        expect(result).toMatchSnapshot()

        rmSync(file)
    })

    testExceptWindows("Range formatting: invalid range values", async () => {
        await mkdir(outputDir, {recursive: true})
        const file = join(outputDir, "range-invalid-values.tolk")
        writeFileSync(file, goodContract)

        // Invalid range values (not numbers)
        const result = await tolkfmt(file, "--range", "a:b-c:d")
        expect(result).toMatchSnapshot()

        rmSync(file)
    })

    testExceptWindows("Range formatting: short form flag", async () => {
        await mkdir(outputDir, {recursive: true})
        const file = join(outputDir, "range-short.tolk")
        const codeWithBadFormatting = `fun foo(){
    val   x   =   1;
    val   y   =   2;
}`
        writeFileSync(file, codeWithBadFormatting)

        // Use short form -r flag
        const result = await tolkfmt(file, "-r", "1:4-1:20")
        expect(result).toMatchSnapshot()

        rmSync(file)
    })

    testExceptWindows("Range formatting: zero-based indexing", async () => {
        await mkdir(outputDir, {recursive: true})
        const file = join(outputDir, "range-zero.tolk")
        const codeWithBadFormatting = `fun foo(){
    val   x   =   1;
}`
        writeFileSync(file, codeWithBadFormatting)

        // Format first line (line 0)
        const result = await tolkfmt(file, "--range", "0:0-0:99")
        expect(result).toMatchSnapshot()

        rmSync(file)
    })

    testExceptWindows("Range formatting: out of bounds", async () => {
        await mkdir(outputDir, {recursive: true})
        const file = join(outputDir, "range-bounds.tolk")
        const shortCode = `fun foo(){}`
        writeFileSync(file, shortCode)

        // Range beyond file bounds, format all file
        const result = await tolkfmt(file, "--range", "0:0-10:50")
        expect(result).toMatchSnapshot()

        rmSync(file)
    })
})
