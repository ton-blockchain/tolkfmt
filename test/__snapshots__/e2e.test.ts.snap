// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`tolkfmt foo.tolk Check and write flags simultaneously 1`] = `
{
  "code": 1,
  "kind": "exited",
  "stderr": "Error: Cannot use both --write and --check options together
",
  "stdout": "",
}
`;

exports[`tolkfmt foo.tolk Check on directory with formatted files 1`] = `
{
  "code": 0,
  "kind": "exited",
  "stderr": "",
  "stdout": "Checking formatting...
All Tolk files are properly formatted!
",
}
`;

exports[`tolkfmt foo.tolk Check on directory with not formatted files 1`] = `
{
  "code": 1,
  "kind": "exited",
  "stderr": "",
  "stdout": "Checking formatting...
[warn] file1.tolk
[warn] file3.tolk
[warn] file2.tolk
Code style issues found in the above files. Run tolkfmt with --write to fix.
",
}
`;

exports[`tolkfmt foo.tolk Check on several directories 1`] = `
{
  "code": 1,
  "kind": "exited",
  "stderr": "",
  "stdout": "Checking formatting...
[warn] file2.tolk
Code style issues found in the above files. Run tolkfmt with --write to fix.
",
}
`;

exports[`tolkfmt foo.tolk Check on several formatted files 1`] = `
{
  "code": 0,
  "kind": "exited",
  "stderr": "",
  "stdout": "Checking formatting...
All Tolk files are properly formatted!
",
}
`;

exports[`tolkfmt foo.tolk Default run 1`] = `
{
  "code": 0,
  "kind": "exited",
  "stderr": "",
  "stdout": "fun test(x: int, y: string): string {
    return x + y;
}
",
}
`;

exports[`tolkfmt foo.tolk Default run with write to file 1`] = `
"fun test(x: int, y: string): string {
    return x + y;
}
"
`;

exports[`tolkfmt foo.tolk Format struct 1`] = `
{
  "code": 0,
  "kind": "exited",
  "stderr": "",
  "stdout": "struct MyStruct {
    x: int = 42
    y: string = "hello"
}
",
}
`;

exports[`tolkfmt foo.tolk Help flag 1`] = `
{
  "code": 0,
  "kind": "exited",
  "stderr": "",
  "stdout": "tolkfmt/0.0.13

Usage:
  $ tolkfmt [options] <files or directories>

Options:
  -v, --version        Display version number 
  -w, --write          Write result to same file 
  -c, --check          Check if the given files are formatted 
  -r, --range <range>  Format only the specified range (format: startLine:startChar-endLine:endChar) 
  -s, --sort-imports   Sort imports in the formatted file 
  -h, --help           Display this message 
",
}
`;

exports[`tolkfmt foo.tolk No arguments 1`] = `
{
  "code": 0,
  "kind": "exited",
  "stderr": "",
  "stdout": "",
}
`;

exports[`tolkfmt foo.tolk Range formatting: basic 1`] = `
{
  "code": 0,
  "kind": "exited",
  "stderr": "",
  "stdout": "fun foo() {
    val x = 1;
    val   y   =   2;
    val   z   =   3;
    return x + y + z;
}
",
}
`;

exports[`tolkfmt foo.tolk Range formatting: invalid format 1`] = `
{
  "code": 1,
  "kind": "exited",
  "stderr": "Error: Invalid range format. Expected: startLine:startChar-endLine:endChar (e.g., 1:5-3:10)
",
  "stdout": "",
}
`;

exports[`tolkfmt foo.tolk Range formatting: invalid range values 1`] = `
{
  "code": 1,
  "kind": "exited",
  "stderr": "Error: Invalid range format. Expected: startLine:startChar-endLine:endChar (e.g., 1:5-3:10)
",
  "stdout": "",
}
`;

exports[`tolkfmt foo.tolk Range formatting: multiple lines 1`] = `
{
  "code": 0,
  "kind": "exited",
  "stderr": "",
  "stdout": "fun foo() {
    val x = 1;
    val y = 2;
    val   z   =   3;
    return x + y + z;
}
",
}
`;

exports[`tolkfmt foo.tolk Range formatting: out of bounds 1`] = `
{
  "code": 0,
  "kind": "exited",
  "stderr": "",
  "stdout": "fun foo() {}
",
}
`;

exports[`tolkfmt foo.tolk Range formatting: short form flag 1`] = `
{
  "code": 0,
  "kind": "exited",
  "stderr": "",
  "stdout": "fun foo() {
    val x = 1;
    val   y   =   2;
}
",
}
`;

exports[`tolkfmt foo.tolk Range formatting: struct 1`] = `
{
  "code": 0,
  "kind": "exited",
  "stderr": "",
  "stdout": "type MyType = int;

struct MyStruct {
    field1: int
    field2: string
}

fun foo() {
    val   x   =   1;
}
",
}
`;

exports[`tolkfmt foo.tolk Range formatting: with check 1`] = `
{
  "code": 1,
  "kind": "exited",
  "stderr": "",
  "stdout": "Checking formatting...
[warn] range-check.tolk
Code style issues found in the above files. Run tolkfmt with --write to fix.
",
}
`;

exports[`tolkfmt foo.tolk Range formatting: zero-based indexing 1`] = `
{
  "code": 0,
  "kind": "exited",
  "stderr": "",
  "stdout": "fun foo() {
    val   x   =   1;
}
",
}
`;

exports[`tolkfmt foo.tolk Run on directory 1`] = `
"fun foo1() {}
"
`;

exports[`tolkfmt foo.tolk Run on directory 2`] = `
"fun foo2() {}
"
`;

exports[`tolkfmt foo.tolk Run on directory 3`] = `
"fun foo3() {}
"
`;

exports[`tolkfmt foo.tolk Version flag 1`] = `"tolkfmt/0.0.13"`;
